internalshared/configutil/normalize.go

```diff
@@ -4,88 +4,106 @@
-// address, the normalized copy will be conformant with RFC-5942 §4
-// See: https://rfc-editor.org/rfc/rfc5952.html
-func NormalizeAddr(address string) string {
-	if address == "" {
+// NormalizeAddr takes a string of a Host, Host:Port, URL, or Destination
+// Address and returns a copy where any IP addresses have been normalized to be
+// conformant with RFC 5942 §4. If the input string does not match any of the
+// supported syntaxes, or the "host" section is not an IP address, the input
+// will be returned unchanged. Supported syntaxes are:
+//
+//	Host                 host                                or [host]
+//	Host:Port            host:port                           or [host]:port
+//	URL                  scheme://user@host/path?query#frag  or scheme://user@[host]/path?query#frag
+//	Destination Address  user@host:port                      or user@[host]:port
+//
+// See:
+//
+//	https://rfc-editor.org/rfc/rfc3986.html
+//	https://rfc-editor.org/rfc/rfc5942.html
+//	https://rfc-editor.org/rfc/rfc5952.html
+func NormalizeAddr(addr string) string {
+	if addr == "" {
 		return ""
 	}
 
-	var ip net.IP
-	var port string
-	bracketedIPv6 := false
+	// Host
+	ip := net.ParseIP(addr)
+	if ip != nil {
+		// net.IP.String() is RFC 5942 §4 compliant
+		return ip.String()
+	}
 
-	// Try parsing it as a URL
-	pu, err := url.Parse(address)
-	if err == nil {
-		// We've been given something that appears to be a URL. See if the hostname
-		// is an IP address
-		ip = net.ParseIP(pu.Hostname())
-	} else {
-		// We haven't been given a URL. Try and parse it as an IP address
-		ip = net.ParseIP(address)
-		if ip == nil {
-			// We haven't been given a URL or IP address, try parsing an IP:Port
-			// combination.
-			idx := strings.LastIndex(address, ":")
-			if idx > 0 {
-				// We've perhaps received an IP:Port address
-				addr := address[:idx]
-				port = address[idx+1:]
-				if strings.HasPrefix(addr, "[") && strings.HasSuffix(addr, "]") {
-					addr = strings.TrimPrefix(strings.TrimSuffix(addr, "]"), "[")
-					bracketedIPv6 = true
-				}
-				ip = net.ParseIP(addr)
-			}
+	// [Host]
+	if strings.HasPrefix(addr, "[") && strings.HasSuffix(addr, "]") {
+		if len(addr) < 3 {
+			return addr
 		}
-	}
 
-	// If our IP is nil whatever was passed in does not contain an IP address.
-	if ip == nil {
-		return address
+		// If we've been given a bracketed IP address, return the address without
+		// normalized without brackets.
+		ip := net.ParseIP(addr[1 : len(addr)-1])
+		if ip != nil {
+			return ip.String()
+		}
+
+		// Our input is not a valid schema.
+		return addr
 	}
 
-	if v4 := ip.To4(); v4 != nil {
-		return address
+	// Host:Port
+	host, port, err := net.SplitHostPort(addr)
+	if err == nil {
+		ip := net.ParseIP(host)
+		if ip == nil {
+			// Our host isn't an IP address so we can return it unchanged
+			return addr
+		}
+
+		// net.JoinHostPort handles bracketing for RFC 5952 §6
+		return net.JoinHostPort(ip.String(), port)
 	}
```

Comment:
Do we need to care about what `err` is in this case? Or is it just important to know when it's nil? If I'm following this correctly, if `net.SplitHostPort()` returns an error, we just move on to `url.Parse()`.

Reply:
Yup, if `net.SplitHostPort()` doesn't work then it does not match the `Host:Port` syntax, so we move on to try another one.